File: forge-core/internal/workintentcontract/validation.go
package workintentcontract

import (
	"bytes"
	"encoding/hex"
	"fmt"
	"regexp"
)

var adr0045Identifier = regexp.MustCompile(`^[a-z0-9][a-z0-9._:/-]{0,159}\z`)

func validateWorkIntentFields(value *WorkIntent, allowBlankIdentity bool) error {
	if value == nil {
		return fmt.Errorf("WorkIntent is nil")
	}
	if err := validateConstants(value); err != nil {
		return err
	}
	if value.DeclaredAtUnixMS < 0 {
		return fmt.Errorf("declared_at_unix_ms must be nonnegative")
	}
	if err := validateAttestations(value.Attestations); err != nil {
		return err
	}
	if err := validateBinding(value.Binding); err != nil {
		return err
	}
	if err := validatePrincipal(value.Requester, "requester"); err != nil {
		return err
	}
	if value.DeclaredOwner != nil {
		if err := validatePrincipal(*value.DeclaredOwner, "declared_owner"); err != nil {
			return err
		}
	}
	if err := validateIntent(value.Intent); err != nil {
		return err
	}
	if err := validateMateriality(value.Materiality); err != nil {
		return err
	}
	if err := validateOrigin(value.Origin); err != nil {
		return err
	}
	if err := validateReferences(value.References); err != nil {
		return err
	}
	return validateIdentityShape(value, allowBlankIdentity)
}

func validateConstants(value *WorkIntent) error {
	expected := []struct{ label, actual, wanted string }{
		{"api_version", value.APIVersion, APIVersion},
		{"canonicalization", value.Canonicalization, Canonicalization},
		{"freshness", value.Freshness, Freshness},
		{"kind", value.Kind, Kind},
		{"status", value.Status, Status},
	}
	for _, item := range expected {
		if item.actual != item.wanted {
			return fmt.Errorf("%s must be %q", item.label, item.wanted)
		}
	}
	return nil
}

func validateAttestations(value Attestations) error {
	values := []bool{value.Approval, value.Authentication, value.Authority,
		value.Completion, value.Effect, value.Execution, value.Freshness,
		value.Materiality, value.Ownership, value.Permission, value.Persistence,
		value.ReferenceResolution, value.Scope, value.Truth}
	for _, attestation := range values {
		if attestation {
			return fmt.Errorf("every WorkIntent attestation must be exactly false")
		}
	}
	return nil
}

func validateBinding(value Binding) error {
	if err := validateText(value.ChangeID, "binding.change_id", maxShortBytes); err != nil {
		return err
	}
	if err := validateText(value.ProjectID, "binding.project_id", maxShortBytes); err != nil {
		return err
	}
	if value.RunID != nil {
		return validateText(*value.RunID, "binding.run_id", maxShortBytes)
	}
	return nil
}

func validatePrincipal(value Principal, label string) error {
	if err := validateText(value.AuthorityDomain, label+".authority_domain", maxShortBytes); err != nil {
		return err
	}
	if err := validateText(value.PrincipalID, label+".principal_id", maxShortBytes); err != nil {
		return err
	}
	return validateEnum(value.PrincipalType, label+".principal_type",
		"agent", "human", "operator", "service")
}

func validateIntent(value Intent) error {
	if value.DeadlineUnixMS != nil && *value.DeadlineUnixMS < 0 {
		return fmt.Errorf("intent.deadline_unix_ms must be nonnegative or null")
	}
	if err := validateText(value.Goal, "intent.goal", maxStringBytes); err != nil {
		return err
	}
	lists := []struct {
		label   string
		values  []string
		minimum int
	}{
		{"intent.external_constraints", value.ExternalConstraints, 0},
		{"intent.non_goals", value.NonGoals, 0},
		{"intent.open_questions", value.OpenQuestions, 0},
		{"intent.scope", value.Scope, 1},
		{"intent.success_signals", value.SuccessSignals, 1},
	}
	total := 0
	for _, list := range lists {
		if err := validateNarrativeList(list.values, list.label, list.minimum); err != nil {
			return err
		}
		total += len(list.values)
	}
	if total > maxNarrativeTotal {
		return fmt.Errorf("intent narrative arrays exceed %d total items", maxNarrativeTotal)
	}
	return validateEnum(value.WorkType, "intent.work_type", "question", "defect",
		"small_change", "feature", "refactor", "migration", "incident_response",
		"architecture_evolution")
}

func validateNarrativeList(values []string, label string, minimum int) error {
	if values == nil || len(values) < minimum || len(values) > maxNarrativeItems {
		return fmt.Errorf("%s must contain %d..%d items", label, minimum, maxNarrativeItems)
	}
	seen := make(map[string]struct{}, len(values))
	for index, value := range values {
		if err := validateText(value, fmt.Sprintf("%s[%d]", label, index), maxStringBytes); err != nil {
			return err
		}
		if _, duplicate := seen[value]; duplicate {
			return fmt.Errorf("%s entries must be exact-string unique", label)
		}
		seen[value] = struct{}{}
	}
	return nil
}

func validateMateriality(value Materiality) error {
	if value.Basis != "caller_declaration_only" {
		return fmt.Errorf("materiality.basis must be caller_declaration_only")
	}
	return validateEnum(value.Level, "materiality.level",
		"materiality_not_bound", "L0", "L1", "L2", "L3", "L4")
}

func validateOrigin(value Origin) error {
	if err := validateEnum(value.OriginKind, "origin.origin_kind", "user_request",
		"runtime_signal", "incident", "technical_debt", "reflection_proposal",
		"operator_request", "other"); err != nil {
		return err
	}
	if value.OriginRef != nil {
		return validateText(*value.OriginRef, "origin.origin_ref", maxReferenceBytes)
	}
	return nil
}

func validateReferences(value References) error {
	if value.ClaimRecordRefs == nil || value.EvidenceRecordRefs == nil ||
		value.LocalArtifactDeclarations == nil {
		return fmt.Errorf("reference arrays must be present arrays, not null")
	}
	if len(value.ClaimRecordRefs)+len(value.EvidenceRecordRefs) > maxCombinedRefs {
		return fmt.Errorf("claim and evidence references exceed %d combined", maxCombinedRefs)
	}
	claims, err := validateRecordRefs(value.ClaimRecordRefs, "claim_record_refs")
	if err != nil {
		return err
	}
	evidence, err := validateRecordRefs(value.EvidenceRecordRefs, "evidence_record_refs")
	if err != nil {
		return err
	}
	for identifier := range claims {
		if _, overlap := evidence[identifier]; overlap {
			return fmt.Errorf("claim and evidence record IDs must be mutually disjoint")
		}
	}
	if err := validateArtifacts(value.LocalArtifactDeclarations); err != nil {
		return err
	}
	return validateSnapshot(value.LocalSourceSnapshotDeclaration)
}

func validateRecordRefs(values []RecordRef, label string) (map[string]struct{}, error) {
	if len(values) > maxRecordRefs {
		return nil, fmt.Errorf("%s exceeds %d items", label, maxRecordRefs)
	}
	seen := make(map[string]struct{}, len(values))
	var previous string
	for index, value := range values {
		if !adr0045Identifier.MatchString(value.RecordID) {
			return nil, fmt.Errorf("%s[%d].record_id violates ADR-0045 grammar", label, index)
		}
		if err := validateHash(value.CanonicalSHA256, label+".canonical_sha256"); err != nil {
			return nil, err
		}
		if index > 0 && bytes.Compare([]byte(previous), []byte(value.RecordID)) >= 0 {
			return nil, fmt.Errorf("%s must be strictly UTF-8 sorted and unique", label)
		}
		seen[value.RecordID] = struct{}{}
		previous = value.RecordID
	}
	return seen, nil
}

func validateArtifacts(values []ArtifactDeclaration) error {
	if len(values) > maxArtifactDecls {
		return fmt.Errorf("local_artifact_declarations exceeds %d items", maxArtifactDecls)
	}
	pairs := make(map[string]struct{}, len(values))
	var previous []byte
	for index, value := range values {
		if err := validateArtifact(value, index); err != nil {
			return err
		}
		pair := value.ArtifactKind + "\x00" + value.ArtifactRef
		if _, duplicate := pairs[pair]; duplicate {
			return fmt.Errorf("artifact kind/ref pairs must be unique")
		}
		pairs[pair] = struct{}{}
		encoded, err := canonicalArtifact(value)
		if err != nil {
			return err
		}
		if index > 0 && bytes.Compare(previous, encoded) >= 0 {
			return fmt.Errorf("artifacts must be strictly sorted by canonical member bytes")
		}
		previous = encoded
	}
	return nil
}

func validateArtifact(value ArtifactDeclaration, index int) error {
	prefix := fmt.Sprintf("local_artifact_declarations[%d]", index)
	if err := validateText(value.ArtifactKind, prefix+".artifact_kind", maxShortBytes); err != nil {
		return err
	}
	if err := validateText(value.ArtifactRef, prefix+".artifact_ref", maxReferenceBytes); err != nil {
		return err
	}
	return validateHash(value.ArtifactSHA256, prefix+".artifact_sha256")
}

func canonicalArtifact(value ArtifactDeclaration) ([]byte, error) {
	return canonicalJSON(map[string]any{
		"artifact_kind": value.ArtifactKind, "artifact_ref": value.ArtifactRef,
		"artifact_sha256": value.ArtifactSHA256,
	})
}

func validateSnapshot(value *SourceSnapshot) error {
	if value == nil {
		return nil
	}
	if !adr0045Identifier.MatchString(value.SnapshotID) {
		return fmt.Errorf("snapshot_id violates ADR-0045 grammar")
	}
	if err := validateHash(value.SnapshotSHA256, "snapshot_sha256"); err != nil {
		return err
	}
	return validateEnum(value.SnapshotType, "snapshot_type",
		"artifact", "external", "repository", "runtime")
}

func validateIdentityShape(value *WorkIntent, allowBlank bool) error {
	if allowBlank && value.WorkIntentID == "" && value.WorkIntentSHA256 == "" {
		return nil
	}
	if err := validateHash(value.WorkIntentSHA256, "work_intent_sha256"); err != nil {
		return err
	}
	if value.WorkIntentID != workIntentIDPrefix+value.WorkIntentSHA256 {
		return fmt.Errorf("work_intent_id must bind work_intent_sha256")
	}
	return nil
}

func validateText(value, label string, maximum int) error {
	if value == "" || len(value) > maximum {
		return fmt.Errorf("%s byte length must be 1..%d", label, maximum)
	}
	if err := validateWireString(value); err != nil {
		return fmt.Errorf("%s: %w", label, err)
	}
	return nil
}

func validateHash(value, label string) error {
	if len(value) != 64 {
		return fmt.Errorf("%s must be 64 lowercase hex characters", label)
	}
	decoded, err := hex.DecodeString(value)
	if err != nil || hex.EncodeToString(decoded) != value {
		return fmt.Errorf("%s must be 64 lowercase hex characters", label)
	}
	return nil
}

func validateEnum(value, label string, allowed ...string) error {
	for _, candidate := range allowed {
		if value == candidate {
			return nil
		}
	}
	return fmt.Errorf("%s has unsupported value %q", label, value)
}

File: forge-runtime/crates/domain/src/work_intent_contract/validation.rs
use std::collections::BTreeSet;

use super::{
    API_VERSION, CANONICALIZATION, LocalArtifactDeclaration, MAX_LOCAL_ARTIFACTS,
    MAX_NARRATIVE_ITEMS, MAX_NARRATIVE_TOTAL, MAX_RECORD_REFS_PER_KIND, MAX_RECORD_REFS_TOTAL,
    MAX_REFERENCE_TEXT_BYTES, MAX_SHORT_TEXT_BYTES, MAX_STRING_BYTES, Principal, RecordReference,
    SourceSnapshot, WorkIntent, WorkIntentAttestations, WorkIntentContractError, invalid, wire,
};

/// Validates one sealed authority-neutral `WorkIntent` declaration.
///
/// # Errors
///
/// Returns an error for invalid fields, bounds, ordering, attestations, or self-identity.
pub fn validate_work_intent(intent: &WorkIntent) -> Result<(), WorkIntentContractError> {
    validate_body(intent)?;
    validate_hash(&intent.work_intent_sha256, "work_intent_sha256")?;
    if intent.work_intent_id != format!("work-intent-{}", intent.work_intent_sha256) {
        return Err(invalid(
            "work_intent_id must equal work-intent- plus work_intent_sha256",
        ));
    }
    wire::canonical_work_intent_unchecked(intent)?;
    let expected = super::codec::work_intent_sha256_unchecked(intent)?;
    if intent.work_intent_sha256 == expected {
        Ok(())
    } else {
        Err(invalid(
            "work_intent_sha256 does not match the canonical preimage",
        ))
    }
}

pub(super) fn validate_body(intent: &WorkIntent) -> Result<(), WorkIntentContractError> {
    if intent.api_version != API_VERSION
        || intent.canonicalization != CANONICALIZATION
        || intent.declared_at_unix_ms < 0
    {
        return Err(invalid(
            "WorkIntent envelope constants or timestamp are invalid",
        ));
    }
    validate_attestations(&intent.attestations)?;
    validate_binding(intent)?;
    if let Some(owner) = &intent.declared_owner {
        validate_principal(owner, "declared_owner")?;
    }
    validate_principal(&intent.requester, "requester")?;
    validate_intent(intent)?;
    validate_materiality_and_origin(intent)?;
    validate_references(intent)
}

fn validate_attestations(
    attestations: &WorkIntentAttestations,
) -> Result<(), WorkIntentContractError> {
    let values = [
        attestations.approval_attestation,
        attestations.authentication_attestation,
        attestations.authority_attestation,
        attestations.completion_attestation,
        attestations.effect_attestation,
        attestations.execution_attestation,
        attestations.freshness_attestation,
        attestations.materiality_attestation,
        attestations.ownership_attestation,
        attestations.permission_attestation,
        attestations.persistence_attestation,
        attestations.reference_resolution_attestation,
        attestations.scope_attestation,
        attestations.truth_attestation,
    ];
    if values.into_iter().any(|value| value) {
        Err(invalid("every WorkIntent attestation must be false"))
    } else {
        Ok(())
    }
}

fn validate_binding(intent: &WorkIntent) -> Result<(), WorkIntentContractError> {
    validate_text(
        &intent.binding.change_id,
        MAX_SHORT_TEXT_BYTES,
        "binding.change_id",
    )?;
    validate_text(
        &intent.binding.project_id,
        MAX_SHORT_TEXT_BYTES,
        "binding.project_id",
    )?;
    if let Some(run_id) = &intent.binding.run_id {
        validate_text(run_id, MAX_SHORT_TEXT_BYTES, "binding.run_id")?;
    }
    Ok(())
}

fn validate_principal(principal: &Principal, label: &str) -> Result<(), WorkIntentContractError> {
    validate_text(
        &principal.authority_domain,
        MAX_SHORT_TEXT_BYTES,
        &format!("{label}.authority_domain"),
    )?;
    validate_text(
        &principal.principal_id,
        MAX_SHORT_TEXT_BYTES,
        &format!("{label}.principal_id"),
    )
}

fn validate_intent(intent: &WorkIntent) -> Result<(), WorkIntentContractError> {
    if intent
        .intent
        .deadline_unix_ms
        .is_some_and(|value| value < 0)
    {
        return Err(invalid(
            "intent.deadline_unix_ms must be nonnegative or null",
        ));
    }
    validate_text(&intent.intent.goal, MAX_STRING_BYTES, "intent.goal")?;
    let declarations = [
        (
            &intent.intent.external_constraints,
            false,
            "external_constraints",
        ),
        (&intent.intent.non_goals, false, "non_goals"),
        (&intent.intent.open_questions, false, "open_questions"),
        (&intent.intent.scope, true, "scope"),
        (&intent.intent.success_signals, true, "success_signals"),
    ];
    let mut total = 0;
    for (values, nonempty, label) in declarations {
        total += validate_narrative(values, nonempty, label)?;
    }
    if total > MAX_NARRATIVE_TOTAL {
        Err(invalid(format!(
            "intent narrative arrays exceed {MAX_NARRATIVE_TOTAL} total items"
        )))
    } else {
        Ok(())
    }
}

fn validate_narrative(
    values: &[String],
    nonempty: bool,
    label: &str,
) -> Result<usize, WorkIntentContractError> {
    if values.len() > MAX_NARRATIVE_ITEMS || (nonempty && values.is_empty()) {
        return Err(invalid(format!("intent.{label} cardinality is invalid")));
    }
    let mut unique = BTreeSet::new();
    for value in values {
        validate_text(value, MAX_STRING_BYTES, &format!("intent.{label}"))?;
        if !unique.insert(value.as_str()) {
            return Err(invalid(format!(
                "intent.{label} must contain unique authored entries"
            )));
        }
    }
    Ok(values.len())
}

fn validate_materiality_and_origin(intent: &WorkIntent) -> Result<(), WorkIntentContractError> {
    if intent.materiality.basis != "caller_declaration_only" {
        return Err(invalid(
            "materiality.basis must equal caller_declaration_only",
        ));
    }
    if let Some(reference) = &intent.origin.origin_ref {
        validate_text(reference, MAX_REFERENCE_TEXT_BYTES, "origin.origin_ref")?;
    }
    Ok(())
}

fn validate_references(intent: &WorkIntent) -> Result<(), WorkIntentContractError> {
    let references = &intent.references;
    validate_record_refs(&references.claim_record_refs, "claim_record_refs")?;
    validate_record_refs(&references.evidence_record_refs, "evidence_record_refs")?;
    if references.claim_record_refs.len() + references.evidence_record_refs.len()
        > MAX_RECORD_REFS_TOTAL
    {
        return Err(invalid(format!(
            "combined record references exceed {MAX_RECORD_REFS_TOTAL}"
        )));
    }
    let claims: BTreeSet<_> = references
        .claim_record_refs
        .iter()
        .map(|reference| reference.record_id.as_str())
        .collect();
    if references
        .evidence_record_refs
        .iter()
        .any(|reference| claims.contains(reference.record_id.as_str()))
    {
        return Err(invalid("claim and evidence record IDs must be disjoint"));
    }
    validate_artifacts(&references.local_artifact_declarations)?;
    if let Some(snapshot) = &references.local_source_snapshot_declaration {
        validate_snapshot(snapshot)?;
    }
    Ok(())
}

fn validate_record_refs(
    references: &[RecordReference],
    label: &str,
) -> Result<(), WorkIntentContractError> {
    if references.len() > MAX_RECORD_REFS_PER_KIND {
        return Err(invalid(format!(
            "{label} exceeds {MAX_RECORD_REFS_PER_KIND} entries"
        )));
    }
    for reference in references {
        validate_identifier(&reference.record_id, &format!("{label}.record_id"))?;
        validate_hash(
            &reference.canonical_sha256,
            &format!("{label}.canonical_sha256"),
        )?;
    }
    if references
        .windows(2)
        .all(|pair| pair[0].record_id.as_bytes() < pair[1].record_id.as_bytes())
    {
        Ok(())
    } else {
        Err(invalid(format!(
            "{label} must be strictly record_id sorted and unique"
        )))
    }
}

fn validate_artifacts(
    artifacts: &[LocalArtifactDeclaration],
) -> Result<(), WorkIntentContractError> {
    if artifacts.len() > MAX_LOCAL_ARTIFACTS {
        return Err(invalid(format!(
            "local artifacts exceed {MAX_LOCAL_ARTIFACTS} entries"
        )));
    }
    let mut pairs = BTreeSet::new();
    let mut encoded = Vec::with_capacity(artifacts.len());
    for artifact in artifacts {
        validate_artifact(artifact)?;
        if !pairs.insert((
            artifact.artifact_kind.as_str(),
            artifact.artifact_ref.as_str(),
        )) {
            return Err(invalid("artifact kind/ref pairs must be unique"));
        }
        let value = serde_json::to_value(artifact)
            .map_err(|error| invalid(format!("artifact cannot be encoded: {error}")))?;
        encoded.push(wire::canonical_json_value(&value)?);
    }
    if encoded.windows(2).all(|pair| pair[0] < pair[1]) {
        Ok(())
    } else {
        Err(invalid(
            "artifacts must be strictly sorted and unique by canonical bytes",
        ))
    }
}

fn validate_artifact(artifact: &LocalArtifactDeclaration) -> Result<(), WorkIntentContractError> {
    validate_text(
        &artifact.artifact_kind,
        MAX_SHORT_TEXT_BYTES,
        "artifact_kind",
    )?;
    validate_text(
        &artifact.artifact_ref,
        MAX_REFERENCE_TEXT_BYTES,
        "artifact_ref",
    )?;
    validate_hash(&artifact.artifact_sha256, "artifact_sha256")
}

fn validate_snapshot(snapshot: &SourceSnapshot) -> Result<(), WorkIntentContractError> {
    validate_identifier(&snapshot.snapshot_id, "snapshot_id")?;
    validate_hash(&snapshot.snapshot_sha256, "snapshot_sha256")
}

fn validate_text(value: &str, maximum: usize, label: &str) -> Result<(), WorkIntentContractError> {
    if value.is_empty() || value.len() > maximum {
        Err(invalid(format!(
            "{label} must contain 1..={maximum} UTF-8 bytes"
        )))
    } else {
        Ok(())
    }
}

fn validate_identifier(value: &str, label: &str) -> Result<(), WorkIntentContractError> {
    let bytes = value.as_bytes();
    let valid = (1..=MAX_SHORT_TEXT_BYTES).contains(&bytes.len())
        && (bytes[0].is_ascii_lowercase() || bytes[0].is_ascii_digit())
        && bytes.iter().all(|byte| {
            byte.is_ascii_lowercase()
                || byte.is_ascii_digit()
                || matches!(*byte, b'.' | b'_' | b':' | b'/' | b'-')
        });
    if valid {
        Ok(())
    } else {
        Err(invalid(format!(
            "{label} must match the ADR-0045 identifier grammar"
        )))
    }
}

fn validate_hash(value: &str, label: &str) -> Result<(), WorkIntentContractError> {
    let valid = value.len() == 64
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte));
    if valid {
        Ok(())
    } else {
        Err(invalid(format!("{label} must be lowercase SHA-256 hex")))
    }
}
